### PR TITLE
ci: fix ignored errors within appraisal loop

### LINKED ISSUE #1097: Can't run tests in Docker Containers
I'm trying to follow [the instructions in the PG Instrumentation specs](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/39056a3190561e87bfa4c5103508cffb0e133713/instrumentation/pg/test/opentelemetry/instrumentation/pg/instrumentation_test.rb#L14-L21) to run the tests, but the tests fail to load `pg` properly:

```
❯ docker-compose run ex-instrumentation-pg-test bundle exec rake test
[+] Creating 1/0
 ✔ Container opentelemetry-ruby-contrib-postgres-1  Running                                                                                         0.0s
/bundle/ruby/3.2.0/gems/activesupport-7.1.3.4/lib/active_support/core_ext/object/blank.rb:121: warning: method redefined; discarding old blank?
/bundle/ruby/3.2.0/gems/rubocop-1.64.1/lib/rubocop/core_ext/string.rb:19: warning: previous definition of blank? was here
<internal:/usr/local/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:37:in `require': cannot load such file -- pg (LoadError)
        from <internal:/usr/local/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:37:in `require'
        from /app/instrumentation/pg/test/opentelemetry/instrumentation/pg/instrumentation_test.rb:9:in `<top (required)>'
        from <internal:/usr/local/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:37:in `require'
        from <internal:/usr/local/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:37:in `require'
        from /bundle/ruby/3.2.0/gems/rake-13.2.1/lib/rake/rake_test_loader.rb:21:in `block in <main>'
        from /bundle/ruby/3.2.0/gems/rake-13.2.1/lib/rake/rake_test_loader.rb:6:in `select'
        from /bundle/ruby/3.2.0/gems/rake-13.2.1/lib/rake/rake_test_loader.rb:6:in `<main>'
rake aborted!
Command failed with status (1)
/bundle/ruby/3.2.0/gems/rake-13.2.1/exe/rake:27:in `<top (required)>'
/bundle/bin/bundle:108:in `load'
/bundle/bin/bundle:108:in `<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

I can see Appraisal in the files for the `pg` instrumentation Gem, but that doesn't seem to be able to install the pg due to some C error (I'm struggling to see it through the screed of warnings that are also displayed.

Am I running the tests wrong? I'm trying to create a new Instrumentation Gem and I'm using PG/Mysql2 as prior art to figure out how to build it, but at the moment I don't see how I'm supposed to be running tests, or at the very least the instructions that are there seem to be either incomplete or incorrect.

### PR DESCRIPTION
**Pull request by robbkidd**
The existing loop over appraisals for a gem is swallowing any error exits returned by commands within the loop.

Resolves #1097. The specific error mentioned in that issue for failure to bundle install pg 1.2 was fixed in #1108. This change fixes the appraisal loop.

Fix appraisal loop swallowing errors:

* add an explicit exit to the appraisal loop if any command in the chain fails so that the CI step also fails

Fix other errors that were being swallowed:

* add a workaround for GEM_HOME file permission problems during bundle install
  *  introduced to GitHub runner images sometime early July 2024
* bound which version of Ruby the gruf gem tests are run against
  * its gemspec includes upper bounds on Ruby versions which cause bundler install to fail under newer Rubies

### CODE REVIEW

**Comment by robbkidd** [comment]
This has unearthed filesystem permissions set in the pre-built runners that bundler does not approve of. There is [an issue open about world-writtable Ruby install location](https://github.com/actions/runner-images/issues/10215). I'll add a workaround in the meantime.

**Comment by robbkidd** [comment]
It might make sense to fix the pg tests in a separate PR depending on how we'd like to communicate the drop support. This PR so far is a non-breaking change.

File: .github/actions/test_gem/action.yml [hunk 1]
```diff
@@ -100,7 +107,7 @@ runs:
             echo "::group::🔎 Appraising ${i}"
             BUNDLE_GEMFILE=gemfiles/${i}.gemfile bundle install --quiet --jobs=3 --retry=4 && \
             BUNDLE_GEMFILE=gemfiles/${i}.gemfile bundle show && \
-            BUNDLE_GEMFILE=gemfiles/${i}.gemfile bundle exec rake test
+            BUNDLE_GEMFILE=gemfiles/${i}.gemfile bundle exec rake test || exit
```
Comment (arielvalentin) [inline .github/actions/test_gem/action.yml:110]: would using `bundle install --path` using a relative vendor directory address the permissions issue?

Comment (robbkidd) [reply .github/actions/test_gem/action.yml:110]: Probably! I cannot confirm a fix locally, so I'll throw another commit on here changing the workaround from the separate `chmod` step to including a `--path` switch here. Then I'll push that up and see how CI reacts.

**Comment by kaylareopelle** [comment]
@robbkidd  - Thanks for putting this together! It looks like the issue about the versions of `pg` we're testing will be fixed in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1106

File: .github/actions/test_gem/action.yml (hunk 1 above)
Comment (robbkidd) [reply .github/actions/test_gem/action.yml:105]: `bundle install --path` is deprecated, so I went with `bundle config path` in d7ba5f4. Now watching CI. ⌛

**Comment by robbkidd** [comment]
@kaylareopelle, to get CI on `main` passing appropriately and _soon_, I am probably going to open a new PR for the change to the pg Appraisals. @matthewtusker [said](https://github.com/open-telemetry/opentelemetry-ruby-contrib/issues/1097#issuecomment-2276047954) it's going to take some time to navigate the CLA signing process with his company and it would be fine with him for someone else to submit the change.

File: .github/actions/test_gem/action.yml (hunk 1 above)
Comment (robbkidd) [reply .github/actions/test_gem/action.yml:105]: Nope. `bundle config path` didn't work for bundler's filesystem permission check. I'm going to revert that commit and go back to `chmod`.

Comment (arielvalentin) [reply .github/actions/test_gem/action.yml:110]: @simi any tips here?

**Pull request merged**